### PR TITLE
Linux Installer

### PR DESCRIPTION
**Pull request by karottenreibe**
Addresses issue [TS-31573](https://cqse.atlassian.net/browse/TS-31573)

- [x] Changes are tested adequately
- [ ] Agent's README.md updated in case of user-visible changes
- [x] CHANGELOG.md updated
- [ ] Present new features in [N&N](https://wiki.cqse.eu/pages/viewpage.action?pageId=689566)
- [ ] [TGA Tutorial](https://docs.teamscale.com/tutorial/setting-up-tga-java) updated
- [ ] [TIA Tutorial](https://docs.teamscale.com/tutorial/tia-java/) updated

Please respect the vote of the [Teamscale bot](https://demo.teamscale.com) or flag irrelevant findings as tolerated or false positives. If you feel that the Teamscale config needs adjustment, please state so in a comment and discuss this with your reviewer.

### CODE REVIEW

**Comment by karottenreibe** [comment]
Short update after a first manual test on a VM:

- some minor edge cases left to fix
- Tomcat sets `ProtectSystem=strict` in its systemd unit file, which makes /opt no longer writable (for log files and coverage files) https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html
- need to handle multiple processes trying to write to the same log file

**Comment by karottenreibe** [comment]
Last manual test was a success, but I had to resolve some pretty big merge conflicts, so redoing the manual test to be sure

File: installer/src/main/java/com/teamscale/profiler/installer/TeamscaleUtils.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+				.header("Authorization", Credentials.basic(credentials.username, credentials.accessKey))
+				.build();
+
+		try (Response response = client.newCall(request).execute()) {
+			handleErrors(response, credentials);
+		} catch (SSLException e) {
+			throw new FatalInstallerError("Failed to connect via HTTPS to " + credentials.url
+					+ "\nPlease ensure that your Teamscale instance is reachable under " + credentials.url
+					+ " and that it is configured for HTTPS, not HTTP. E.g. open that URL in your"
+					+ " browser and verify that you can connect successfully.",
```
Comment (DreierF) [inline installer/src/main/java/com/teamscale/profiler/installer/TeamscaleUtils.java:41]: Might also happen if the server uses an internally signed SSL certificate that is not installed on the current machine, right?

File: installer/src/test/java/com/teamscale/profiler/installer/MockTeamscale.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import spark.Response;
+import spark.Service;
+
+import static javax.servlet.http.HttpServletResponse.SC_INTERNAL_SERVER_ERROR;
+import static javax.servlet.http.HttpServletResponse.SC_OK;
+
+/**
+ * Mocks a redirect server: redirects all requests to another server.
+ */
+	public class MockTeamscale {
```
Comment (DreierF) [inline installer/src/test/java/com/teamscale/profiler/installer/MockTeamscale.java:13]: Formatting off

File: system-tests/debug-logging-default-test/build.gradle.kts [hunk 3]
```diff
@@ -1,7 +1,15 @@
...
 
 tasks.test {
-    teamscaleAgent(mapOf("debug" to "true"))
+	teamscaleAgent(mapOf("debug" to "true"))
+
+	doFirst {
+		// the test checks if a log file is written, so we need to remove all leftover agent temp directories
+		// to avoid false-positive test failures
+		val leftoverAgentTempDirectories = file(System.getProperty("java.io.tmpdir")).listFiles()
+			.filter { it.name.contains("teamscale-java-profiler") }
```
Comment (DreierF) [inline system-tests/debug-logging-default-test/build.gradle.kts:12]: The system tests are running in parallel. Doesn't this become a problem if another system test writes stuff to there?

File: system-tests/teamscale-properties-test/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,34 @@
+import kotlin.io.path.writeText
+
+plugins {
+	com.teamscale.`system-test-convention`
+}
+
+tasks.test {
+	/** These ports must match what is configured in the SystemTest class. */
+	val agentPort = 63301
```
Comment (DreierF) [inline system-tests/teamscale-properties-test/build.gradle.kts:9]: This agent and Teamscale port is already used by HttpRedirectSystemTest. Please pick a different one to not cause issues when executing the tests in parallel

**Review by DreierF — CHANGES_REQUESTED** [review]

File: agent/build.gradle.kts [hunk 5]
```diff
@@ -12,13 +12,15 @@ plugins {
 	com.teamscale.`logger-patch`
 }
 
+evaluationDependsOn(":installer")
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/build.gradle.kts:15]: I didn't think I'd see the day when I teach you something about Gradle :P :partying_face: ;-) (though I must admit I just googled this one)

File: system-tests/debug-logging-default-test/build.gradle.kts (hunk 3 above)
Comment (karottenreibe) [reply system-tests/debug-logging-default-test/build.gradle.kts:12]: ah, good point. Didn't think about this. I now changed the test to not use `/tmp` but a private directory as the system temp dir

File: system-tests/teamscale-properties-test/build.gradle.kts (hunk 4 above)
Comment (karottenreibe) [reply system-tests/teamscale-properties-test/build.gradle.kts]: good point, I wasn't aware that we need to avoid conflicts here. I augmented the system test convention so ports are now picked automatically so they don't conflict. I hope I got this right ;-)

File: installer/src/main/java/com/teamscale/profiler/installer/InstallCommand.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * CLI command that installs the profiler.
+ */
+@CommandLine.Command(name = "install", description = "Installs the profiler.")
+public class InstallCommand implements Callable<Integer> {
+
+	@CommandLine.Parameters(index = "0", converter = HttpUrlTypeConverter.class, description = "The URL of your Teamscale instance.")
+	private HttpUrl teamscaleUrl;
+
+	@CommandLine.Parameters(index = "1", description = "The user used to access your Teamscale instance and upload coverage to your projects.")
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply installer/src/main/java/com/teamscale/profiler/installer/InstallCommand.java:19]: tbh during the installation it can be any valid user. We can only check if the user can access Teamscale. Since during installation we don't have any particular project yet that we know should be accessed, we can't check permissions at this point other than "can authenticate with Teamscale".

File: installer/src/main/java/com/teamscale/profiler/installer/Installer.java [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+/** Installs the agent system-globally. */
+public class Installer {
+
+	private static final Path DEFAULT_INSTALL_DIRECTORY = Paths.get("/opt/teamscale-profiler/java");
+	private static final Path DEFAULT_ETC_DIRECTORY = Paths.get("/etc");
+
+	private static final String RERUN_ADVICE = getRerunAdvice();
+
+	private static String getRerunAdvice() {
+		if (SystemUtils.isWindows()) {
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply installer/src/main/java/com/teamscale/profiler/installer/Installer.java:27]: ah yes, I got carried away here and started thinking about how/if this could work on windows :P do you want me to remove it or leave it in for later when we implement Windows support? It's sort of dead code ATM, so can't be triggered (since we don't compile for Windows)

[the remark this answers is not in the document]
Comment (DreierF) [reply installer/src/main/java/com/teamscale/profiler/installer/Installer.java:27]: You can leave it. Was just confused :D

File: agent/src/main/java/com/teamscale/jacoco/agent/util/AgentUtils.java [hunk 8]
```diff
@@ -2,14 +2,55 @@
...
+	 */
+	public static String getPID() {
+		try {
+			Class<?> processHandleClass = Class.forName("java.lang.ProcessHandle");
+			Object processHandle = processHandleClass.getMethod("current").invoke(null);
+			Long pid = (Long) processHandleClass.getMethod("pid").invoke(processHandle);
+			return pid.toString();
+		} catch (ClassNotFoundException | NoSuchMethodException | IllegalAccessException |
+				 InvocationTargetException e) {
+			return ManagementFactory.getRuntimeMXBean().getName();
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/src/main/java/com/teamscale/jacoco/agent/util/AgentUtils.java:36]: it might. I'm sanitizing this now

File: agent/src/test/java/com/teamscale/jacoco/agent/options/AgentOptionsParserTest.java [hunk 9]
```diff
@@ -3,16 +3,40 @@
...
-	private final AgentOptionsParser parser = new AgentOptionsParser(new CommandLineLogger());
+	private final AgentOptionsParser parser = new AgentOptionsParser(new CommandLineLogger(), null, null);
+
+	@Test
+	public void environmentConfigOverridesCommandLineOptions(
+			@TempDir Path tempDir) throws IOException, AgentOptionParseException {
+		Path config = tempDir.resolve("config.properties");
+		Files.write(config, "debug=true".getBytes(StandardCharsets.UTF_8));
+		AgentOptionsParser parser = new AgentOptionsParser(new CommandLineLogger(), config.toString(), null);
+		AgentOptions options = parser.parse("debug=false");
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/src/test/java/com/teamscale/jacoco/agent/options/AgentOptionsParserTest.java:29]: we could, but this would be a bigger change to how the option parsing works since that doesnt keep track of how often an option occurs yet. So I'll create a ticket for that instead. I'll document that env options override command line options to at least make it clear in the docs

File: agent/src/main/java/com/teamscale/jacoco/agent/PreMain.java [hunk 10]
```diff
@@ -31,11 +34,32 @@ public class PreMain {
...
 	private static LoggingUtils.LoggingResources loggingResources = null;
 
+	/**
+	 * System property that we use to prevent this agent from being attached to the same VM twice. This can happen if
+	 * the agent is registered via multiple JVM environment variables and/or the command line at the same time.
+	 */
+	private static final String LOCKING_SYSTEM_PROPERTY = "TEAMSCALE_JAVA_PROFILER_ATTACHED";
+
+	/** Environment variable from which to read the config file to use. */
+	private static final String CONFIG_ENVIRONMENT_VARIABLE = "TEAMSCALE_JAVA_PROFILER_CONFIG";
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/src/main/java/com/teamscale/jacoco/agent/PreMain.java:44]: Moved to https://cqse.atlassian.net/browse/TS-36383

File: agent/src/test/java/com/teamscale/jacoco/agent/options/AgentOptionsParserTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (karottenreibe) [reply agent/src/test/java/com/teamscale/jacoco/agent/options/AgentOptionsParserTest.java:29]: Created https://cqse.atlassian.net/browse/TS-36384 and moved documentation task to https://cqse.atlassian.net/browse/TS-36383

File: installer/build.gradle.kts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+application {
+	applicationName = "installer"
+	mainClass.set("com.teamscale.profiler.installer.RootCommand")
+}
+
+dependencies {
+	// we need this older version since newer versions are Kotlin-implemented and don't play nice with GraalVM
+	implementation("com.squareup.okhttp3:okhttp:3.14.2")
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply installer/build.gradle.kts:21]: No, official support will come with okhttp 5 which is currently alpha. Added a link to the changelog instead :-)

File: installer/src/main/java/com/teamscale/profiler/installer/Installer.java [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+		try {
+			getDefaultInstaller().runInstall(credentials);
+			System.out.println("Installation successful. Profiler installed to " + DEFAULT_INSTALL_DIRECTORY);
+			if (SystemUtils.isLinux()) {
+				System.out.println("To use the profiler in your current session, please log out and back in.");
+			}
+			System.out.println("To activate the profiler for an application, set the environment variable:"
+					+ "\nTEAMSCALE_JAVA_PROFILER_CONFIG"
+					+ "\nIts value must be the path to a valid profiler configuration file."
+					+ "\nThen, restart your application (for web applications: restart the app server).");
```
[the remark this answers is not in the document]
Comment (karottenreibe) [reply installer/src/main/java/com/teamscale/profiler/installer/Installer.java:83]: Moved to https://cqse.atlassian.net/browse/TS-36383

File: buildSrc/src/main/kotlin/SystemTestUtils.kt [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+			val pickedPort = nextFreePort
+			nextFreePort += 1
+			return pickedPort
+		}
+	}
+
+}
+
+/** The port that the fake Teamscale should use during the system test. Guaranteed conflict-free. */
+val Test.teamscalePort: Int by lazy { SystemTestUtils.pickFreePort() }
```
Comment (DreierF) [inline buildSrc/src/main/kotlin/SystemTestUtils.kt:28]: Very nice idea!
Unfortunately this won't work reliably, because Gradle already uses classloader isolation for subprojects that do not apply the exact same set of plugins, which is already the case for some system test modules. This will also be used more in the future with configuration caching and project isolation being enabled. The only reliable solution for this problem is to make `SystemTestUtils` a [shared build service](https://docs.gradle.org/current/userguide/build_services.html) (see also the related discussion on [Github](https://github.com/gradle/gradle/issues/14697))

File: installer/build.gradle.kts [hunk 14]
```diff
@@ -18,6 +18,7 @@ application {
 
 dependencies {
 	// we need this older version since newer versions are Kotlin-implemented and don't play nice with GraalVM
+	// okhttp 5 will add official GraalVM support but is currently still alpha https://square.github.io/okhttp/changelogs/changelog/
 	implementation("com.squareup.okhttp3:okhttp:3.14.2")
```
Comment (DreierF) [inline installer/build.gradle.kts:22]: You can also add this to the `libs.versions.toml` and add add the following to the `renovate.json`:
```
    {
      "matchPackageNames": [  "com.squareup.okhttp3:okhttp" ],
      "allowedVersions": ">= 5"
    },
```
Then we'll get a PR once 5.0 is available and we cannot miss it :)

**Review by DreierF — CHANGES_REQUESTED** [review]

File: buildSrc/src/main/kotlin/SystemTestUtils.kt (hunk 13 above)
Comment (karottenreibe) [reply buildSrc/src/main/kotlin/SystemTestUtils.kt:28]: I did my best but I'm not 100% sure I implemented this build service right. The docs are all about using the service in custom tasks but we're modifying the existing Test task, so that doesn't seem to apply. At least in my local setup, it picks all distinct ports, but I don't think it ran anything in parallel anyways, so better review this extra carefully ;-)

File: installer/build.gradle.kts (hunk 14 above)
Comment (karottenreibe) [reply installer/build.gradle.kts:22]: hm, not sure this is so easy since the agent already uses a 4.* version of okhttp. So this would mean downgrading the agent's library, right? Which probably means changing a bunch of code since 3 and 4 are not API compatible, 4 being a complete rewrite in Kotlin.

Comment (DreierF) [reply installer/build.gradle.kts:22]: Ah sorry, I missed that we already use v4.

File: buildSrc/src/main/kotlin/SystemTestPorts.kt [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+}
+
+private val teamscalePortMap = mutableMapOf<Test, Int>()
+
+/** The port that the fake Teamscale should use during the system test. Guaranteed conflict-free. */
+val Test.teamscalePort: Int
+	get() {
+		synchronized(teamscalePortMap) {
+			return teamscalePortMap.getOrPut(this) { SystemTestPorts.pickFreePort() }
```
Comment (DreierF) [inline buildSrc/src/main/kotlin/SystemTestPorts.kt]: This might work, but I'm not really sure to be honest as we are now sharing one instance of the Provider potentially across projects. I would suggest the following:

Store the chosen provider and the chosen port as an extension property directly on the test instance so that we don't need to keep the test instances in a map, which kind of feels dangerous.
```
var Test.portProvider: Provider<SystemTestPorts>
	get() = extra["portProvider"] as Provider<SystemTestPorts>
	set(value) {
		extra["portProvider"] = value
	}

var Test.teamscalePort: Int
	get() = extra["teamscalePort"] as Int
	set(value) {
		extra["teamscalePort"] = value
	}
```

In `com.teamscale.system-test-convention.gradle.kts` you can do:
```
val provider = SystemTestPorts.registerWith(project)

tasks.test {
//...
	usesService(provider)
	portProvider = provider
	teamscalePort = provider.get().pickFreePort()
	agentPort = provider.get().pickFreePort()
	systemProperties("agentPort" to agentPort, "teamscalePort" to teamscalePort)
}
```
and hence don't have to rely on the global provider instance anymore.

For the redirect/artifactory ports it becomes `val artifactoryPort = portProvider.get().pickFreePort()`

I committed the changes to your branch as I already had them locally anyway. Please have look if you agree: https://github.com/cqse/teamscale-jacoco-agent/pull/287/commits/fa72d4daa0131ec810d24f8e98eabc918587170d

**Review by DreierF — CHANGES_REQUESTED** [review]
Thanks for the rework! I pushed two commits one to disable the InstallerTests for non linux systems and the other one to refactor the Gradle code (see separate comment).

However there is still a problem with the test setup. When I run `./gradlew build` locally I get 
```
> Task :system-tests:artifactory-git-properties-detection:test FAILED
Failed to parse agent options: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
Exception in thread "main" java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
        at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
Caused by: com.teamscale.jacoco.agent.options.AgentOptionParseException: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
        at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:100)
        at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:67)
FATAL ERROR in native method: processing of -javaagent failed
        at com.teamscale.jacoco.agent.PreMain.getAndApplyAgentOptions(PreMain.java:92)
        at com.teamscale.jacoco.agent.PreMain.premain(PreMain.java:62)
```

This is because the test picks up the `teamscale.properties` file from a previous test run.
I guess this validation error could also become a problem if you would want to configure the agent to upload to artifactory in general.

File: buildSrc/src/main/kotlin/SystemTestPorts.kt (hunk 15 above)
Comment (karottenreibe) [reply buildSrc/src/main/kotlin/SystemTestPorts.kt]: looks excellent, thank you!

**Comment by karottenreibe** [comment]
> Thanks for the rework! I pushed two commits one to disable the InstallerTests for non linux systems and the other one to refactor the Gradle code (see separate comment).
> 
> However there is still a problem with the test setup. When I run `./gradlew build` locally I get
> 
> ```
> > Task :system-tests:artifactory-git-properties-detection:test FAILED
> Failed to parse agent options: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
> Exception in thread "main" java.lang.reflect.InvocationTargetException
>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>         at java.lang.reflect.Method.invoke(Method.java:498)
>         at sun.instrument.InstrumentationImpl.loadClassAndStartAgent(InstrumentationImpl.java:386)
>         at sun.instrument.InstrumentationImpl.loadClassAndCallPremain(InstrumentationImpl.java:401)
> Caused by: com.teamscale.jacoco.agent.options.AgentOptionParseException: Invalid options given: - You did provide some options prefixed with 'teamscale-', but not all required ones!
>         at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:100)
>         at com.teamscale.jacoco.agent.options.AgentOptionsParser.parse(AgentOptionsParser.java:67)
> FATAL ERROR in native method: processing of -javaagent failed
>         at com.teamscale.jacoco.agent.PreMain.getAndApplyAgentOptions(PreMain.java:92)
>         at com.teamscale.jacoco.agent.PreMain.premain(PreMain.java:62)
> ```
> 
> This is because the test picks up the `teamscale.properties` file from a previous test run. I guess this validation error could also become a problem if you would want to configure the agent to upload to artifactory in general.

good catch. I now isolated the system tests so each of them uses their own copy of the shadowed agent jar, avoiding this problem

**Comment by DreierF** [comment]
> good catch. I now isolated the system tests so each of them uses their own copy of the shadowed agent jar, avoiding this problem

But that means you'd still not be able to install the agent via the installer and upload to artifactory, right?

**Comment by karottenreibe** [comment]
Correct, I haven't gotten to that part yet 😋

Florian Dreier ***@***.***> schrieb am Fr., 17. Nov. 2023,
17:47:

**Comment by karottenreibe** [comment]
I fixed this now, but that was a bigger operation. To make sure I don't accidentally break anything, I added a bunch of additional tests for the option parsing for valid and invalid combinations.
I also refactored some of the option parsing code to enable my changes and make it easier to understand.
I also simplified some of the NWDI parsing code since that was unnecessarily complicated and also had a bug that the additional tests revealed.

Sorry for the big code change but it wasn't really possible otherwise AFAICS.

**Comment by karottenreibe** [comment]
There's one architecture finding for which we IMO need to adjust the architecture to allow the dependency. So leaving that as is.

**Review by DreierF — APPROVED** [review]
Thanks for the cleanup! Really appreciate it!
